File: Editorial/0101-0200/0120_Triangle/JavaScript/4_Memoization.js
let triangle;
let memoTable = new Map();
function minPath(row, col) {
    let params = row + ":" + col;
    if (memoTable.has(params)) {
        return memoTable.get(params);
    }
    let path = triangle[row][col];
    if (row < triangle.length - 1) {
        path += Math.min(minPath(row + 1, col), minPath(row + 1, col + 1));
    }
    memoTable.set(params, path);
    return path;
}
var minimumTotal = function (triangleInput) {
    triangle = triangleInput;
    memoTable.clear();
    return minPath(0, 0);
};

File: Editorial/0101-0200/0120_Triangle/CPP/4_Memoization.cpp
class Solution {
private:
    vector<vector<int>> triangle;
    map<pair<int, int>, int> memoTable;
    int minPath(int row, int col) {
        if (memoTable.count({row, col})) {
            return memoTable[{row, col}];
        }
        int path = triangle[row][col];
        if (row < triangle.size() - 1) {
            path += min(minPath(row + 1, col), minPath(row + 1, col + 1));
        }
        memoTable[{row, col}] = path;
        return path;
    }

public:
    int minimumTotal(vector<vector<int>>& triangle) {
        this->triangle = triangle;
        return minPath(0, 0);
    }
};
